### PR TITLE
Feature/#187update student profile

### PR DESCRIPTION
**Pull request by diegdar**
It can update student profile for the fields: name, surname, linkedin_url, github_url, about and tags_ids for technologies development from the endpoint: /api/v1/student/{studentId}/resume/profile

The feature uses validations in the backend to ensure that the fields meet the requirements  expected when the endpoint request is received and tests to ensure that both the functionalities of the service methods and those of the controller do not generate bugs.

To test that the changes were made correctly you can see them on the endpoint:
/api/v1/student/{studentId}/resume/detail

### CODE REVIEW

File: app/Service/Student/UpdateStudentProfileService.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+};
+
+class UpdateStudentProfileService
+{
+    public function execute(string $studentId, array $data): bool
+    {
+        return $this->updateStudentProfile($data, $studentId);
+    }
+
+    public function updateStudentProfile(array $data, string $studentId):bool
```
Comment (jordimorillo) [inline app/Service/Student/UpdateStudentProfileService.php:19]: change the order of the parameters (first the subject, then the action)

File: app/Service/Student/UpdateStudentProfileService.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        $student = $this->getStudentById($studentId);
+        $wasStudentUPdated = $this->updateStudent($data, $student);
+        $wasResumeUpdated = $this->updateResume($data, $student);
```
Comment (jordimorillo) [inline app/Service/Student/UpdateStudentProfileService.php:23]: The return of this method should be void as it is a command. Control the Exceptions to show incomplete actions

File: app/Service/Student/UpdateStudentProfileService.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    private function getStudentById(string $studentId): Student
+    {
+        $student = Student::find($studentId);
+        if (!$student) throw new StudentNotFoundException($studentId);
+
+        return $student;
+    }
+
+    private function updateStudent(array $data, Student $student):bool
```
Comment (jordimorillo) [inline app/Service/Student/UpdateStudentProfileService.php:36]: subject first then the action (in the parameters)

File: app/Service/Student/UpdateStudentProfileService.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        return $student->update([
+            'name' => $data['name'],
+            'surname' => $data['surname'],
+        ]);
+    }
+
+    private function updateResume(array $data, Student $student):bool
```
Comment (jordimorillo) [inline app/Service/Student/UpdateStudentProfileService.php:44]: subject first:

private function updateResume(Student $student, array $data):bool

File: app/Service/Student/UpdateStudentProfileService.php [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        $resume =  $student->resume;
+        if (!$resume) throw new ResumeNotFoundException($student->id);
```
Comment (jordimorillo) [inline app/Service/Student/UpdateStudentProfileService.php:47]: always use the claudators:

if (!$resume) {
     throw new ResumeNotFoundException($student->id);
}

**Review by jordimorillo — APPROVED** [review]
Before merging, fix the issue related to missing fields to update.

File: app/Http/Controllers/api/Student/UpdateStudentProfileController.php [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+        DB::beginTransaction();
+        try {
+            $this->updateStudentProfileService->execute($studentId, $dataStudentProfileUpdate);
+            DB::commit();
+
+                return response()->json([
+                    'profile'=> 'El perfil del estudiante se actualizo correctamente'
+                ], 200);
+        } catch (\DomainException $e) {
```
[the remark this answers is not in the document]
Comment (diegdar) [reply app/Http/Controllers/api/Student/UpdateStudentProfileController.php:38]: done

File: app/Service/Student/UpdateStudentProfileService.php (hunk 1 above)
Comment (diegdar) [reply app/Service/Student/UpdateStudentProfileService.php:19]: done

File: app/Service/Student/UpdateStudentProfileService.php (hunk 2 above)
Comment (diegdar) [reply app/Service/Student/UpdateStudentProfileService.php:23]: done

File: app/Service/Student/UpdateStudentProfileService.php (hunk 3 above)
Comment (diegdar) [reply app/Service/Student/UpdateStudentProfileService.php:36]: done

File: app/Service/Student/UpdateStudentProfileService.php (hunk 4 above)
Comment (diegdar) [reply app/Service/Student/UpdateStudentProfileService.php:44]: done

File: app/Service/Student/UpdateStudentProfileService.php (hunk 5 above)
Comment (diegdar) [reply app/Service/Student/UpdateStudentProfileService.php:47]: done

File: tests/Feature/Service/Student/UpdateStudentProfileServiceTest.php [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    public function test_can_instantiate_student_update_profile_service()
+    {
+        $this->assertNotNull($this->updateStudentProfileService);
+    }
+
+    public function test_can_update_student_profile()
+    {
+        $student = Student::factory()->has(Resume::factory())->create();
+        $dataToUpdate = $this->createFakeDataToUpdate($student);
+        $result = $this->updateStudentProfileService->execute($dataToUpdate['id'], $dataToUpdate);
```
[the remark this answers is not in the document]
Comment (diegdar) [reply tests/Feature/Service/Student/UpdateStudentProfileServiceTest.php:51]: done

File: tests/Feature/Service/Student/UpdateStudentProfileServiceTest.php [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    public function test_update_student_profile_throws_resume_not_found_exception()
+    {
+        $this->expectException(ResumeNotFoundException::class);
+
+        $student = Student::factory()->create();
+        $dataToUpdate = $this->createFakeDataToUpdate($student);
+
+        try {
```
[the remark this answers is not in the document]
Comment (diegdar) [reply tests/Feature/Service/Student/UpdateStudentProfileServiceTest.php:89]: done

File: tests/Feature/Service/Student/UpdateStudentProfileServiceTest.php [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    /**
+     * @dataProvider missingDataProvider
+     */
+    public function test_can_not_update_student_profile_with_missing_data(array $invalidData)
+    {
+        $student = Student::factory()->has(Resume::factory())->create();
+        $dataToUpdate = $this->createFakeDataToUpdate($student);
+        $dataToUpdate = array_merge($dataToUpdate, $invalidData);
+
+        try {
```
[the remark this answers is not in the document]
Comment (diegdar) [reply tests/Feature/Service/Student/UpdateStudentProfileServiceTest.php:106]: done

File: tests/Feature/Controller/Student/UpdateStudentProfileControllerTest.php [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+    public function test_can_update_student_profile(): void
+    {
+        $dataToUpdate = $this->createFakeDataToUpdate();
+        $response = $this->json('PUT', 'api/v1/student/' . $dataToUpdate['id'] . '/resume/profile', $dataToUpdate);
+
+        $response->assertStatus(200);
+        $response->assertJson(['profile' => 'El perfil del estudiante se actualizo correctamente']);
+
+        $this->assertDatabaseHas('students', [
```
[the remark this answers is not in the document]
Comment (diegdar) [reply tests/Feature/Controller/Student/UpdateStudentProfileControllerTest.php:51]: done

File: app/Http/Requests/UpdateStudentRequest.php [hunk 11]
```diff
@@ -26,15 +28,11 @@ public function rules(): array
         return [
             'name' => 'required|string|regex:/^([^0-9]*)$/',
             'surname' => 'required|string|regex:/^([^0-9]*)$/',
-            'email' => 'required|string|email|max:255|unique:users',
             'subtitle' => 'required|string',
-            'bootcamp' => 'required|in:Front end Developer,PHP Developer,Java Developer,Nodejs Developer',
+            'github_url' => 'string|url|max:60|nullable',
```
[the remark this answers is not in the document]
Comment (diegdar) [reply app/Http/Requests/UpdateStudentRequest.php:32]: done

**Comment by JustTheDude001** [comment]
It seems all correct. All tests passed, the changes in the database are done and can be seen using the specified endpoint:
GET /student/{studentId}/resume/detail.
Just one note, a description for the PR would be welcome! Great job!
(And maybe for the future instead of ids for the tasks would be better with the name of the tags, I do not really know, another doubt to ask to Jordi)
Congrats for the job done!!

**Comment by JustTheDude001** [comment]
It seems all correct. All tests passed, the changes in the database are done and can be seen using the specified endpoint:
GET /student/{studentId}/resume/detail.
Just one note, a description for the PR would be welcome! Great job!
(And maybe for the future instead of ids for the tasks would be better with the name of the tags, I do not really know, another doubt to ask to Jordi)
Congrats for the job done!!

**Review by JustTheDude001 — APPROVED** [review]
It seems all correct. All tests passed, the changes in the database are done and can be seen using the specified endpoint:
GET /student/{studentId}/resume/detail.
Just one note, a description for the PR would be welcome! Great job!
(And maybe for the future instead of ids for the tasks would be better with the name of the tags, I do not really know, another doubt to ask to Jordi)
Congrats for the job done!!

**Pull request merged**